core/claims/claim.go

```diff
@@ -131,35 +111,187 @@ func NewClaimFromEntry(e *merkletree.Entry) (merkletree.Entrier, error) {
-	case *ClaimTypeEthId:
-		c := NewClaimEthIdFromEntry(e)
-		return c, nil
-	case *ClaimTypeAuthEthKey:
-		c := NewClaimAuthEthKeyFromEntry(e)
-		return c, nil
+	// case *ClaimTypeSetRootKey:
+	// 	c := NewClaimSetRootKeyFromEntry(e)
+	// 	return c, nil
+	// case *ClaimTypeAuthorizeKSignSecp256k1:
+	// 	return NewClaimAuthorizeKSignSecp256k1FromEntry(e)
+	// case *ClaimTypeLinkObjectIdentity:
+	// 	c := NewClaimLinkObjectIdentityFromEntry(e)
+	// 	return c, nil
+	// case *ClaimTypeAuthorizeService:
+	// 	c := NewClaimAuthorizeServiceFromEntry(e)
+	// 	return c, nil
+	// case *ClaimTypeEthId:
+	// 	c := NewClaimEthIdFromEntry(e)
+	// 	return c, nil
+	// case *ClaimTypeAuthEthKey:
+	// 	c := NewClaimAuthEthKeyFromEntry(e)
+	// 	return c, nil
 	default:
 		return nil, ErrInvalidClaimType
 	}
 }
+
+// ClaimRecip is the flag option to specify a recipient of a claim
+type ClaimRecip byte
+
+const (
+	// ClaimRecipSelf is a claim that refers to a property of the issuing
+	// identity.
+	ClaimRecipSelf ClaimRecip = 0b00
+	// ClaimRecipIdenIndex is a claim that refers to a property of an
+	// identity found in the index part of the claim.
+	ClaimRecipIdenIndex ClaimRecip = 0b01
+	// ClaimRecipIdenIndex is a claim that refers to a property of an
+	// identity found in the value part of the claim.
+	ClaimRecipIdenValue ClaimRecip = 0b10
+)
+
+// ClaimHeader represents the first bytes of the claim index and contains its
+// type and flags.
+type ClaimHeader struct {
+	Type       ClaimType
+	Dest       ClaimRecip
+	Expiration bool
+	Version    bool
+}
+
+func bool2byte(b bool) byte {
+	if b {
+		return 1
+	}
+	return 0
+}
+
+func byte2bool(b byte) bool {
+	return b != 0
+}
+
+// Marshal the ClaimHeader into an entry
+func (c *ClaimHeader) Marshal(e *merkletree.Entry) {
+	index := e.Index()
+	copy(index[0][:ClaimTypeLen], c.Type[:])
+	flags0 := &index[0][ClaimTypeLen]
+	*flags0 = 0
+	*flags0 |= byte(c.Dest)
+	*flags0 |= bool2byte(c.Expiration) << 2
+	*flags0 |= bool2byte(c.Version) << 3
+}
+
+// Unmarshal the ClaimHeader from an entry
+func (c *ClaimHeader) Unmarshal(e *merkletree.Entry) {
+	index := e.Index()
+	copy(c.Type[:], index[0][:ClaimTypeLen])
+	flags0 := index[0][ClaimTypeLen]
+	c.Dest = ClaimRecip(flags0 & 0b00000011)
+	c.Expiration = byte2bool(flags0 & (1 << 2))
+	c.Version = byte2bool(flags0 & (1 << 3))
+}
+
+var (
+	// ClaimHeaderBasic is a simple claim type that can be used for anything.
+	ClaimHeaderBasic = ClaimHeader{
+		Type:       *ClaimTypeBasic,
+		Dest:       ClaimRecipSelf,
+		Expiration: false,
+		Version:    false}
+	// ClaimTypeAuthorizeKSignBabyJub is a claim type to autorize a babyjub public key for signing.
+	ClaimHeaderAuthorizeKSignBabyJub = ClaimHeader{
+		Type:       *ClaimTypeAuthorizeKSignBabyJub,
+		Dest:       ClaimRecipSelf,
+		Expiration: false,
+		Version:    false}
+)
+
+// EntrierMetadata is an intefrace that extends Entrier with a function that
+// returns the claim metadata.
+type EntrierMetadata interface {
```

Comment:
Proposal: `EntrierMetadata` --> `Claimer`.
The idea is that a `Claimer` interface is what all the `Claims` are satisfying. So the interface don't get a name like `EntrierMetadata`, when is a superset of `Entrier`, and is more natural as superset as the `Claimer` contain a `Entrier`.